<a href="https://colab.research.google.com/github/Rajat-Sri/Python-Coding-Practice/blob/main/Python_Coding_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic Operations In Python



![](https://i.imgur.com/gvSnw4A.png)



## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter.


In [ ]:
2+3+9

14

In [ ]:
99-73

26

In [ ]:
34*-56.7

-1927.8000000000002

In [ ]:
3018/14

215.57142857142858

In [ ]:
100//7

14

In [ ]:
5**3

125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


## Variables:
 While working with a programming language such as Python, information is stored in variables. You can think of variables as containers for storing data. The data stored within a variable is called its value.

In [ ]:
a = 4
b = 5
c = a+b
c

9

 **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines.

Inline and single-line comments start with `#`, whereas multi-line comments begin and end with three quotes, i.e. `"""`. Here are some examples of code comments:

In [ ]:
my_favorite_number = 1 # an inline comment

In [ ]:
# This comment gets its own line
my_least_favorite_number = 3

In [ ]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5

## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.

In [ ]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [ ]:
# Equality check - True
my_favorite_number == 1

True

In [ ]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [ ]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [ ]:
# Not equal check - False
a_neutral_number != 3

False

In [ ]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [ ]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [ ]:
# Less than check - False
my_least_favorite_number < my_favorite_number

False

In [ ]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [ ]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

False

Just like arithmetic operations, the result of a comparison operation can also be stored in a variable.

In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |

In [ ]:
my_favorite_number

1

In [ ]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [ ]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [ ]:
True and False

False

In [ ]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [ ]:
a_neutral_number = 3

In [ ]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [ ]:
my_favorite_number < 0 or True

True

The `not` operator returns `False` if a condition is `True` and `True` if the condition is `False`.

In [ ]:
not a_neutral_number == 3

False

In [ ]:
not my_favorite_number < 0

True

Logical operators can be combined to form complex conditions. Use round brackets or parentheses `(` and `)` to indicate the order in which logical operators should be applied.

In [ ]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [ ]:
not (True and 0 < 1) or (False and True)

False

If parentheses are not used, logical operators are applied from left to right.

In [ ]:
not True and 0 < 1 or False and True

False

# Variables and Data Types

A Quick Tour of Variables and Data Types in Python


![](https://i.imgur.com/6cg2E9Q.png)


Storing information using variables

In [ ]:
my_favorite_color = "blue"

In [ ]:
my_favorite_color

'blue'

You can also assign values to multiple variables in a single statement by separating the variable names and values with commas.

In [ ]:
color1, color2, color3 = "red", "green", "blue"

In [ ]:
color1

'red'

In [ ]:
color3

'blue'

You can assign the same value to multiple variables by chaining multiple assignment operations within a single statement.

In [ ]:
color4 = color5 = color6 = "magenta"

In [ ]:
color5

'magenta'

In [ ]:
color6

'magenta'

You can change the value stored within a variable by assigning a new value to it using another assignment statement. Be careful while reassigning variables: when you assign a new value to the variable, the old value is lost and no longer accessible.

In [ ]:
my_favorite_color

'blue'

In [ ]:
my_favorite_color = "red"

In [ ]:
my_favorite_color

'red'

While reassigning a variable, you can also use the variable's previous value to compute the new value.

In [ ]:
counter = 10

In [ ]:
counter = counter + 1

In [ ]:
counter

11

Python provides a shorthand syntax for it.

In [ ]:
# Same as `counter = counter + 4`
counter += 4

In [ ]:
counter

15

Variable names can be short (`a`, `x`, `y`, etc.) or descriptive ( `my_favorite_color`, `profit_margin`, `the_3_musketeers`, etc.). However, you must follow these rules while naming Python variables:

* A variable's name must start with a letter or the underscore character `_`. It cannot begin with a number.
* A variable name can only contain lowercase (small) or uppercase (capital) letters, digits, or underscores (`a`-`z`, `A`-`Z`, `0`-`9`, and `_`).
* Variable names are case-sensitive, i.e., `a_variable`, `A_Variable`, and `A_VARIABLE` are all different variables.

Here are some valid variable names:

In [ ]:
a_variable = 23
is_today_Saturday = False
my_favorite_car = "Delorean"
the_3_musketeers = ["Athos", "Porthos", "Aramis"]

Let's try creating some variables with invalid names. Python prints a syntax error if your variable's name is invalid

In [ ]:
a variable = 23

SyntaxError: ignored

In [ ]:
is_today_$aturday = False

SyntaxError: ignored

In [ ]:
my-favorite-car = "Delorean"

SyntaxError: ignored

In [ ]:
3_musketeers = ["Athos", "Porthos", "Aramis"]

SyntaxError: ignored

## Built-in data types in Python

Any data or information stored within a Python variable has a *type*. You can view the type of data stored within a variable using the `type` function.

In [ ]:
a_variable

23

In [ ]:
type(a_variable)

int

In [ ]:
is_today_Saturday

False

In [ ]:
type(is_today_Saturday)

bool

In [ ]:
the_3_musketeers

['Athos', 'Porthos', 'Aramis']

In [ ]:
type(the_3_musketeers)

list

Booleans are automatically converted to `int`s when used in arithmetic operations. `True` is converted to `1` and `False` is converted to `0`.

In [ ]:
5 + False

5

In [ ]:
3.0 + True

4.0

Any value in Python can be converted to a Boolean using the `bool` function.

Only the following values evaluate to `False` (they are often called *falsy* values):

1. The value `False` itself
2. The integer `0`
3. The float `0.0`
4. The empty value `None`
5. The empty text `""`
6. The empty list `[]`
7. The empty tuple `()`
8. The empty dictionary `{}`
9. The empty set `set()`
10. The empty range `range(0)`

Everything else evaluates to `True` (a value that evaluates to `True` is often called a *truthy* value).

In [ ]:
bool(True), bool(1), bool(2.0), bool("hello"), bool([1,2]), bool((2,3)), bool(range(10))

(True, True, True, True, True, True, True)

In [ ]:
bool(False), bool(0), bool(0.0), bool(None), bool(""), bool([]), bool(()), bool({}), bool(set()), bool(range(0))

(False, False, False, False, False, False, False, False, False, False)

### None

The None type includes a single value `None`, used to indicate the absence of a value. `None` has the type `NoneType`. It is often used to declare a variable whose value may be assigned later.

In [ ]:
nothing = None

In [ ]:
type(nothing)

NoneType

You can use single quotes inside a string written with double quotes, and vice versa.

In [ ]:
my_favorite_movie = "One Flew over the Cuckoo's Nest"

In [ ]:
my_favorite_movie

"One Flew over the Cuckoo's Nest"

In [ ]:
my_favorite_pun = 'Thanks for explaining the word "many" to me, it means a lot.'

In [ ]:
my_favorite_pun

'Thanks for explaining the word "many" to me, it means a lot.'

To create multiline strings, use three single quotes ''' or three double quotes """ to begin and end the string. Line breaks are represented using the newline character \n

In [ ]:
yet_another_pun = '''Son: "Dad, can you tell me what a solar eclipse is?"
Dad: "No sun."'''

In [ ]:
yet_another_pun

'Son: "Dad, can you tell me what a solar eclipse is?" \nDad: "No sun."'

Multiline strings are best displayed using the `print` function.

In [ ]:
print(yet_another_pun)

Son: "Dad, can you tell me what a solar eclipse is?" 
Dad: "No sun."


You can check the length of a string using the `len` function.

In [ ]:
len(my_favorite_movie)

31

A string can be converted into a list of characters using `list` function.

In [ ]:
today = 'saturday'

In [ ]:
list(today)

['s', 'a', 't', 'u', 'r', 'd', 'a', 'y']

You can access individual characters within a string using the [] indexing notation.

In [ ]:
today[3]

'u'

In [ ]:
today[-1]

'y'

You can access a part of a string using by providing a `start:end` range instead of a single index in `[]`.

In [ ]:
today[2:4]

'tu'

You can also check whether a string contains a some text using the `in` operator.

In [ ]:
'a' in today

True

In [ ]:
'z' in today

False

Two or more strings can be joined or concatenated using the + operator. Be careful while concatenating strings, sometimes you may need to add a space character " " between words.

In [ ]:
full_name = "Derek O'Brien"

In [ ]:
greeting = "Hello"

In [ ]:
greeting + full_name

"HelloDerek O'Brien"

In [ ]:
greeting + " " + full_name + "!" # additional space

"Hello Derek O'Brien!"

Strings in Python have many built-in *methods* that are used to manipulate them. Let's try out some common string methods.

> **Methods**: Methods are functions associated with data types and are accessed using the `.` notation e.g. `variable_name.method()` or `"a string".method()`. Methods are a powerful technique for associating common operations with values of specific data types.

The `.lower()`, `.upper()` and `.capitalize()` methods are used to change the case of the characters.

In [ ]:
today = "Saturday"
today

'Saturday'

In [ ]:
today.lower()

'saturday'

In [ ]:
"saturday".upper()

'SATURDAY'

In [ ]:
"monday".capitalize() # changes first character to uppercase

'Monday'

The `.replace` method replaces a part of the string with another string. It takes the portion to be replaced and the replacement text as *inputs* or *arguments*.

In [ ]:
another_day = today.replace("Satur", "Wednes")

In [ ]:
another_day

'Wednesday'

Note that `replace` returns a new string, and the original string is not modified.

In [ ]:
today

'Saturday'

The `.split` method splits a string into a list of strings at every occurrence of provided character(s).

In [ ]:
"Sun,Mon,Tue,Wed,Thu,Fri,Sat".split(",")

['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']

The `.strip` method removes whitespace characters from the beginning and end of a string.

In [ ]:
a_long_line = "       This is a long line with some space before, after,     and some space in the middle..    "

In [ ]:
a_long_line_stripped = a_long_line.strip()

In [ ]:
a_long_line_stripped

'This is a long line with some space before, after,     and some space in the middle..'

The `.format` method combines values of other data types, e.g., integers, floats, booleans, lists, etc. with strings. You can use `format` to construct output messages for display.

In [ ]:
# Input variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Template for output message
output_template = """If a grocery store sells ice bags at $ {} per bag, with a profit margin of {} %,
then the total profit it makes by selling {} ice bags is $ {}."""

print(output_template)

If a grocery store sells ice bags at $ {} per bag, with a profit margin of {} %, 
then the total profit it makes by selling {} ice bags is $ {}.


In [ ]:
# Inserting values into the string
total_profit = cost_of_ice_bag * profit_margin * number_of_bags
output_message = output_template.format(cost_of_ice_bag, profit_margin*100, number_of_bags, total_profit)

print(output_message)

If a grocery store sells ice bags at $ 1.25 per bag, with a profit margin of 20.0 %, 
then the total profit it makes by selling 500 ice bags is $ 125.0.


You can `str` to convert a value of any data type into a string.

In [ ]:
str(23),str(23.432),str(True)

('23', '23.432', 'True')

In [ ]:
the_3_musketeers = ["Athos", "Porthos", "Aramis"]
str(the_3_musketeers)

"['Athos', 'Porthos', 'Aramis']"

## List

A list in Python is an ordered collection of values. Lists can hold values of different data types and support operations to add, remove, and change values. Lists have the type `list`.

To create a list, enclose a sequence of values within square brackets `[` and `]`, separated by commas.

In [ ]:
fruits = ['apple', 'banana', 'cherry']

In [ ]:
fruits

['apple', 'banana', 'cherry']

In [ ]:
type(fruits)

list

Let's try creating a list containing values of different data types, including another list.

In [ ]:
a_list = [23, 'hello', None, 3.14, fruits, 3 <= 5]

In [ ]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

To determine the number of values in a list, use the `len` function. You can use `len`  to determine the number of values in several other data types.

In [ ]:
len(fruits)

3

In [ ]:
print("Number of fruits:", len(fruits))

Number of fruits: 3


You can access an element from the list using its *index*, e.g., `fruits[2]` returns the element at index 2 within the list `fruits`. The starting index of a list is 0.

In [ ]:
fruits[0]

'apple'

In [ ]:
fruits[1]

'banana'

In [ ]:
fruits[3]

IndexError: ignored

In [ ]:
fruits[-1]

'cherry'

You can also access a range of values from the list. The result is itself a list. Let us look at some examples.

In [ ]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [ ]:
len(a_list)

6

In [ ]:
a_list[2:5]

[None, 3.14, ['apple', 'banana', 'cherry']]

You can also change the value at a specific index within a list using the assignment operation.

In [ ]:
fruits

['apple', 'banana', 'cherry']

In [ ]:
fruits[1] = 'blueberry'

In [ ]:
fruits

['apple', 'blueberry', 'cherry']

A new value can be added to the end of a list using the `append` method.

In [ ]:
fruits.append('dates')

In [ ]:
fruits

['apple', 'blueberry', 'cherry', 'dates']

You can remove a value from a list using the `remove` method.

In [ ]:
fruits.remove('blueberry')

In [ ]:
fruits

['apple', 'cherry', 'dates']


If a list has multiple instances of the value passed to the .remove() method, only the first occurrence of that value will be removed from the list.

In [ ]:
my_list = [1, 2, 3, 2, 4, 2, 5]

# Removing the value 2
my_list.remove(2)

In [ ]:
my_list

[1, 3, 2, 4, 2, 5]

To remove an element from a specific index, use the `pop` method. The method also returns the removed element.

In [ ]:
fruits

['apple', 'cherry', 'dates']

In [ ]:
fruits.pop(1)

'cherry'

In [ ]:
fruits

['apple', 'dates']

If no index is provided, the `pop` method removes the last element of the list.

In [ ]:
fruits.pop()

'dates'

In [ ]:
fruits

['apple']

You can test whether a list contains a value using the `in` operator.

In [ ]:
'pineapple' in fruits

False

In [ ]:
'apple' in fruits

True

To combine two or more lists, use the `+` operator. This operation is also called *concatenation*.

In [ ]:
more_fruits = fruits + ['pineapple', 'tomato', 'guava'] + ['dates', 'banana']

In [ ]:
more_fruits

['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']

To create a copy of a list, use the `copy` method. Modifying the copied list does not affect the original.

In [ ]:
more_fruits_copy = more_fruits.copy()

In [ ]:
more_fruits_copy

['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']

In [ ]:
# Modify the copy
more_fruits_copy.remove('pineapple')
more_fruits_copy.pop()
more_fruits_copy

['apple', 'tomato', 'guava', 'dates']

In [ ]:
# Original list remains unchanged
more_fruits

['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']

Note that you cannot create a copy of a list by simply creating a new variable using the assignment operator =. The new variable will point to the same list, and any modifications performed using either variable will affect the other.

Reverse the order of elements in a list:

In [ ]:
more_fruits = ['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']
more_fruits.reverse()
print(more_fruits)

['banana', 'dates', 'guava', 'tomato', 'pineapple', 'apple']


Add the elements of one list at the end of another list:

In [ ]:
more_fruits = ['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']
additional_fruits = ['orange', 'kiwi', 'mango']
more_fruits.extend(additional_fruits)
print(more_fruits)


['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana', 'orange', 'kiwi', 'mango']


Sort a list of strings in alphabetical order:

In [ ]:
more_fruits = ['apple', 'pineapple', 'tomato', 'guava', 'dates', 'banana']
more_fruits.sort()
print(more_fruits)


['apple', 'banana', 'dates', 'guava', 'pineapple', 'tomato']


Sort a list of numbers in decreasing order:

In [ ]:
numbers = [5, 10, 2, 8, 3]
numbers.sort(reverse=True)
print(numbers)

[10, 8, 5, 3, 2]


## Tuple

A tuple is an ordered collection of values, similar to a list. However, it is not possible to add, remove, or modify values in a tuple. A tuple is created by enclosing values within parentheses `(` and `)`, separated by commas.

> Any data structure that cannot be modified after creation is called *immutable*. You can think of tuples as immutable lists.

Let's try some experiments with tuples.

In [2]:
fruits = ('apple', 'cherry', 'dates')

In [4]:
# check no. of elements
len(fruits)

3

In [5]:
# get an element (positive index)
fruits[0]

'apple'

You can also skip the parantheses `(` and `)` while creating a tuple. Python automatically converts comma-separated values into a tuple.

In [6]:
the_3_musketeers = 'Athos', 'Porthos', 'Aramis'

In [7]:
the_3_musketeers

('Athos', 'Porthos', 'Aramis')

In [10]:
single_element_tuple = 4,
another_single_element_tuple = (4,)
not_a_tuple = (4)

In [11]:
print(single_element_tuple,another_single_element_tuple,not_a_tuple )

(4,) (4,) 4


You can convert a list into a tuple using the `tuple` function, and vice versa using the `list` function

In [12]:
tuple(['one', 'two', 'three'])

('one', 'two', 'three')

In [13]:
list(('Athos', 'Porthos', 'Aramis'))

['Athos', 'Porthos', 'Aramis']

Tuples have just two built-in methods: count and index.

### Dictionary

A dictionary is an unordered collection of items. Each item stored in a dictionary has a key and value. You can use a key to retrieve the corresponding value from the dictionary.  Dictionaries have the type `dict`.

Dictionaries are often used to store many pieces of information e.g. details about a person, in a single variable. Dictionaries are created by enclosing key-value pairs within braces or curly brackets `{` and `}`.

In [14]:
person1 = {
    'name': 'John Doe',
    'sex': 'Male',
    'age': 32,
    'married': True
}

In [15]:
person1

{'name': 'John Doe', 'sex': 'Male', 'age': 32, 'married': True}

Dictionaries can also be created using the `dict` function.

In [16]:
person2 = dict(name='Jane Judy', sex='Female', age=28, married=False)

In [17]:
person2

{'name': 'Jane Judy', 'sex': 'Female', 'age': 28, 'married': False}

Keys can be used to access values using square brackets `[` and `]`.

In [18]:
person1['name']

'John Doe'

You can also use the `get` method to access the value associated with a key.

In [19]:
person2.get("name")

'Jane Judy'

You can check whether a key is present in a dictionary using the `in` operator.

In [20]:
'address' in person1

False

You can change the value associated with a key using the assignment operator.

In [21]:
person2['married']

False

In [22]:
person2['married'] = True

In [23]:
person2['married']

True

The assignment operator can also be used to add new key-value pairs to the dictionary.